### PR TITLE
network: Limit message length based on Tag

### PR DESCRIPTION
**Pull request by iansuvak**
## Summary

This PR implements specific message length limits based on the protocol Tag of the message being currently read in the `wsPeer`. 

The new `<TypeName>MaxSize() int` methods are being generated by the `msgp` library based on the supplied `allocbound`s and the new struct annotation `totalallocbound` used to limit the total number of bytes encoded in a collection type. 

New constants are defined and `Tag.MaxSize()` method switching between them is used when reseting `LimitedReaderSlurper`. These don't override the global maxAllocation setting so if max message size is larger than the maxAllocation we will still error out with the old limit. 

TODOs:
- [x] Fix allocbounds/totalallocbounds on StateProofs that are currently blowing up the maxsize of transactions. 
- [x] ~Write a method calculating maxSize for `TS` `TopicMsgRes` tags~ We decided to keep `TS` and `VB` at existing maximum message size since their theoretical limits weren't meaningfully different. 
- [x] ~Implement actual enforcement of limits on topic messages (MI, TS, UE) since those are not read using `LimitedReadSlurper`~ This statement was false. They are read using the slurper.

## Test Plan

- [x] Add new tests to make sure that constants are defined for each Tag
- [x] Add tests to make sure that custom calculated constants are correct
- [x] Write new tests for the slurper to make sure that the per message limits work

### CODE REVIEW

File: protocol/txntype.go [hunk 1]
```diff
@@ -20,8 +20,11 @@ package protocol
 // in a block.  They are used in the data/transaction package and the REST API.
 
 // TxType is the type of the transaction written to the ledger
+//msgp:allocbound TxType txTypeMaxLen
 type TxType string
 
+const txTypeMaxLen = 7
```
Comment (algorandskiy) [inline protocol/txntype.go:26]: add a test checking all txTypeMaxLen against all possible TxType constants with AST

File: catchup/universalFetcher_test.go [hunk 2]
```diff
@@ -312,3 +313,18 @@ func TestErrorTypes(t *testing.T) {
...
+func TestMaxBlockRequestSize(t *testing.T) {
+	partitiontest.PartitionTest(t)
+
+	round := rand.Uint64()
+	topics := MakeBlockRequestTopics(basics.Round(round))
+	nonce := rand.Uint64() - 1
+	nonceTopic := network.MakeNonceTopic(nonce)
+	topics = append(topics, nonceTopic)
+	serializedMsg := topics.MarshallTopics()
+	require.Equal(t, uint64(len(serializedMsg)), protocol.UniEnsBlockReqTag.MaxMessageSize())
```
[the remark this answers is not in the document]
Comment (iansuvak) [reply catchup/universalFetcher_test.go:328]: Our handrolled encoding uses constant size encoding so it should be exactly the same regardless of the random number rolled but I can change it

File: protocol/txntype.go (hunk 1 above)
Comment (iansuvak) [reply protocol/txntype.go:26]: I don't think that we need to do this since this would break only if we added a new txtype with longer name. If we did that we would be writing tests of making sure the tests work properly and none of those would pass since msgp would fail to unmarshall messages exceeding this allocbound

File: network/limited_reader_slurper.go [hunk 3]
```diff
@@ -110,12 +122,18 @@ func (s *LimitedReaderSlurper) Size() (size uint64) {
...
+func (s *LimitedReaderSlurper) Reset(n uint64) {
 	for i := 1; i <= s.lastBuffer; i++ {
 		s.remainedUnallocatedSpace += uint64(cap(s.buffers[i]))
 		s.buffers[i] = nil
 	}
+	if n > 0 {
+		s.currentMessageMaxSize = n
+	} else {
+		s.currentMessageMaxSize = s.remainedUnallocatedSpace + uint64(cap(s.buffers[0]))
+	}
```
[the remark this answers is not in the document]
Comment (iansuvak) [reply network/limited_reader_slurper.go]: Done

File: node/node_test.go [hunk 4]
```diff
@@ -539,3 +544,31 @@ func TestOfflineOnlineClosedBitStatus(t *testing.T) {
 		})
 	}
 }
+
+func TestMaxSizesCorrect(t *testing.T) {
+	partitiontest.PartitionTest(t)
+
+	avSize := uint64(agreement.UnauthenticatedVoteMaxSize())
```
[the remark this answers is not in the document]
Comment (iansuvak) [reply node/node_test.go:554]: Implemented in `protocol/tags_test.go`

File: agreement/types.go [hunk 5]
```diff
@@ -54,6 +54,11 @@ type (
 	period uint64
 )
 
+// RoundMaxSize is defined manually since it is expected by the msgp generated MaxSize methods.
+func RoundMaxSize() int {
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply agreement/types.go]: removed

File: ledger/testing/randomAccounts.go [hunk 6]
```diff
@@ -100,7 +100,7 @@ func RandomAssetParams() basics.AssetParams {
 		DefaultFrozen: crypto.RandUint64()%2 == 0,
 	}
 	if crypto.RandUint64()%5 != 0 {
-		ap.UnitName = fmt.Sprintf("un%x", uint32(crypto.RandUint64()%0x7fffffff))
+		ap.UnitName = fmt.Sprintf("un%x", uint32(crypto.RandUint64()%0x7fffff))
```
Comment (iansuvak) [inline ledger/testing/randomAccounts.go:103]: This is necessary because otherwise it generates a string that's 10 characters long rather than the 8 allowed by the `MaxAssetUnitNameBytes`

File: agreement/msgp_gen.go [hunk 7]
```diff
@@ -6563,7 +6869,7 @@ func (z *rootRouter) UnmarshalMsg(bts []byte) (o []byte, err error) {
 				(*z).Children = make(map[round]*roundRouter, zb0009)
 			}
 			for zb0009 > 0 {
-				var zb0001 round
+				var zb0001 basics.Round
```
[the remark this answers is not in the document]
Comment (iansuvak) [reply agreement/msgp_gen.go:6872]: That's actually a great question. We should verify this

File: crypto/stateproof/structs.go [hunk 8]
```diff
@@ -82,6 +82,10 @@ type StateProof struct {
 	PositionsToReveal []uint64          `codec:"pr,allocbound=MaxReveals"`
 }
 
+// SigPartProofMaxSize is the maximum valid size of SigProofs and PartProofs elements of the Stateproof struct in bytes.
+// It is equal to merklearray.ProofMaxSizeByElements(config.StateProofTopVoters/2)
```
Comment (algonautshant) [inline crypto/stateproof/structs.go:86]: What is the reason for having this `config.StateProofTopVoters/2` and not `config.StateProofTopVoters`?

File: crypto/stateproof/msgp_gen.go [hunk 9]
```diff
@@ -1409,6 +1458,26 @@ func (z *StateProof) MsgIsZero() bool {
...
 }
 
+// MaxSize returns a maximum valid message size for this message type
+func StateProofMaxSize() (s int) {
+	s = 1 + 2 + crypto.GenericDigestMaxSize() + 2 + msgp.Uint64Size + 2
+	// Using totalallocbound for: z.SigProofs
+	s += SigPartProofMaxSize
+	s += 2
+	// Using totalallocbound for: z.PartProofs
+	s += SigPartProofMaxSize
```
Comment (algonautshant) [inline crypto/stateproof/msgp_gen.go:1468]: How is this generated? We have `s += SigPartProofMaxSize` twice.

File: crypto/stateproof/weights_test.go [hunk 10]
```diff
@@ -184,6 +186,14 @@ func TestNumReveals(t *testing.T) {
 	}
 }
 
+func TestSigPartProofMaxSize(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+	// Ensures that the SigPartProofMaxSize constant used for totalallocbound for StateProof.(Sig|Part)Proof(s) is
+	// correct. It should be logically bound by the maximum number of StateProofTopVoters. It is scaled by 1/2
```
Comment (algonautshant) [inline crypto/stateproof/weights_test.go:193]: Why is it scaled by 1/2?

File: crypto/stateproof/structs.go (hunk 8 above)
Comment (algorandskiy) [reply crypto/stateproof/structs.go:86]: look at `merklearray.Proof` definition and a comment there

>	// Path is bounded by MaxNumLeavesOnEncodedTree since there could be multiple reveals, and
	// given the distribution of the elt positions and the depth of the tree,
	// the path length can increase up to 2^MaxEncodedTreeDepth / 2
	Path        []crypto.GenericDigest `codec:"pth,allocbound=MaxNumLeavesOnEncodedTree/2"`

File: crypto/stateproof/structs.go [hunk 11]
```diff
@@ -82,6 +82,10 @@ type StateProof struct {
...
+const SigPartProofMaxSize = 35353
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply crypto/stateproof/structs.go:87]: agree, need to unify

File: crypto/stateproof/msgp_gen.go (hunk 9 above)
Comment (iansuvak) [reply crypto/stateproof/msgp_gen.go:1468]: Both SigProofs and PartProofs have the same theoretical upper bound so are using the same allocbound constant in `StateProof` struct definition here: https://github.com/iansuvak/go-algorand/blob/3289d45fd12d78481d93f0b0000d0e461fbcc52e/crypto/stateproof/structs.go#L75-L76

File: crypto/merklearray/proof.go [hunk 12]
```diff
@@ -41,12 +42,29 @@ type Proof struct {
...
+	s = 1 + 4
+	// Calculating size of slice: z.Path
+	s += msgp.ArrayHeaderSize + n*(crypto.GenericDigestMaxSize())
+	s += 4 + crypto.HashFactoryMaxSize() + 3 + msgp.Uint8Size
+	return
+}
+
+// SingleLeafProofMaxSize returns the maximum msgp encoded size of proof verifying a single leaf
+func SingleLeafProofMaxSize() int {
+	return ProofMaxSizeByElements(MaxEncodedTreeDepth)
```
[the remark this answers is not in the document]
Comment (iansuvak) [reply crypto/merklearray/proof.go:65]: I can make the change but none of these MaxSize() functions are used at runtime outside of testing. The only thing that this PR uses at runtime is constants populating the MaxSize() method on the protocol.Tag type.

File: crypto/stateproof/structs.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (iansuvak) [reply crypto/stateproof/structs.go:87]: To me the cases seem different enough to warrant leaving separate.

SingleLeafProof is a valid type and right now it's only used in `merklesignature.Signature` but if it were used elsewhere the code would either:
 - stop compiling if we kept `//msgp:ignore maxsize SingleLeafProof` 
 - if we removed it, it would compile and give absurdly large values by calculating the size of the whole Proof with no limit to the number of elements. 

SigProof and PartProof on the other hands are just fields and not special type so the constant value totalallocbound was the only way of limiting their size.

File: crypto/merklearray/proof.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (algonautshant) [reply crypto/merklearray/proof.go:65]: > none of these MaxSize() functions are used at runtime outside of testing.

Then these functions should be defined in the test files, not in production files.

File: crypto/stateproof/msgp_gen.go (hunk 9 above)
Comment (algonautshant) [reply crypto/stateproof/msgp_gen.go:1468]: My bad. Thanks. You very elegantly have also generated the comments.

File: crypto/stateproof/weights_test.go (hunk 10 above)
Comment (iansuvak) [reply crypto/stateproof/weights_test.go:193]: This comes from Idan's explanation and you perhaps understand it better but my understanding is that this is the worst case scenario but even so you only need one side of the tree for each of the voters.

File: crypto/msgp_gen.go [hunk 13]
```diff
@@ -3012,6 +3309,13 @@ func (z *ed25519PrivateKey) MsgIsZero() bool {
 	return (*z) == (ed25519PrivateKey{})
 }
 
+// MaxSize returns a maximum valid message size for this message type
+func Ed25519PrivateKeyMaxSize() (s int) {
+	// Calculating size of array: z
+	s += msgp.ArrayHeaderSize + ((64) * (msgp.ByteSize))
```
[the remark this answers is not in the document]
Comment (iansuvak) [reply crypto/msgp_gen.go:3315]: It was. I just pushed a fix to the msgp repo. It should be reflected once we cut a new version

File: crypto/stateproof/weights_test.go (hunk 10 above)
Comment (algonautshant) [reply crypto/stateproof/weights_test.go:193]: It is from my explanation [here](https://github.com/algorand/go-algorand/pull/3171/files)  :-D
It is not one side of the tree. It is more complicated than that. 

The hints are the values needed to compute the root given the reveals (leaves with `^` under them in the illustration below).
The idea is the following: 

```
	// Path is bounded by MaxNumLeaves since there could be multiple reveals, and
	// given the distribution of the elt positions and the depth of the tree,
	// the path length can increase up to 2^MaxTreeDepth / 2

Consider two different reveals for the same tree:
               z5
       z3              z4
   y       z       z1      z2
 q   r   s   t   u   v   w   x
a b c d e f g h i j k l m n o p
  ^
  hints: [a, r, z, z4]
len(hints) = 4
You need a to combine with b to get q, need r to combine with the computed q and get y, and so on.

The worst case is this:
               z5
       z3              z4
   y       z       z1      z2
 q   r   s   t   u   v   w   x
a b c d e f g h i j k l m n o p
^   ^     ^   ^ ^   ^     ^   ^

hints: [b, d, e, g, j, l, m, o]
len(hints) = 2^4/2
```

File: config/config.go [hunk 14]
```diff
@@ -81,6 +81,14 @@ const ConfigurableConsensusProtocolsFilename = "consensus.json"
 // do not expose in normal config so it is not in code generated local_defaults.go
 const defaultRelayGossipFanout = 8
 
+// MaxGenesisIDLen is the maximum length of the genesis ID set for purpose of setting
+// allocbounds on structs containing GenesisID and for purposes of calculating MaxSize functions
+// on those types. Current value is larger than the existing network IDs and the ones used in testing
+const MaxGenesisIDLen = 128
```
[the remark this answers is not in the document]
Comment (iansuvak) [reply config/config.go:87]: It should never be necessary. The reason why this is so large is as I mentioned in the comment that there are tests that depend on the test name being passed around as the genesisID

File: crypto/stateproof/structs.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (iansuvak) [reply crypto/stateproof/structs.go:87]: Some more thoughts:
We can't add `totalallocbound/maxtotalbytes` to `SingleLeafProof` directly at definition since `Proof` is an embedded field and current msgp implementation doesn't accept annotation on those. 

The `ProofMaxSizeByElements` function has 3 distinct inputs:

- Max size matching full potential `Proof`(unused)
- SigProof and PartProof size depending on the StateProofTopVoters constant
- SingleLeafProof which is dependent on the MaxEncodedTreeDepth

File: agreement/msgp_gen.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (iansuvak) [reply agreement/msgp_gen.go:6872]: With rootRouteBasic defined as 
```
type rootRouterBasic struct {
	_struct struct{} `codec:","`

	root         actor    // playerMachine   (not restored: explicitly set on construction)
	proposalRoot listener // proposalMachine
	voteRoot     listener // voteMachine

	ProposalManager proposalManager
	VoteAggregator  voteAggregator

	Children map[basics.Round]*roundRouter `codec:"Children,allocbound=-"`
}
```

i.e. exactly the same but using basics.Round as the key instead of the alias the following test passes. 

```
func TestMsgpTypeAliasCompat(t *testing.T) {
	partitiontest.PartitionTest(t)

	randomRound := rand.Uint64()
	a := rootRouter{Children: map[round]*roundRouter{round(randomRound): {}}}

	b := rootRouterBasic{Children: map[basics.Round]*roundRouter{basics.Round(randomRound): {}}}

	require.NotEqual(t, a, b)

	aEnc := protocol.Encode(&a)
	bEnc := protocol.Encode(&b)

	require.Equal(t, aEnc, bEnc)

	var aDecA, bDecA rootRouter
	var aDecB, bDecB rootRouterBasic

	err := protocol.Decode(aEnc, &aDecA)
	require.NoError(t, err)
	err = protocol.Decode(bEnc, &bDecA)
	require.NoError(t, err)
	require.Equal(t, aDecA, bDecA)

	err = protocol.Decode(aEnc, &aDecB)
	require.NoError(t, err)
	err = protocol.Decode(bEnc, &bDecB)
	require.NoError(t, err)
	require.Equal(t, aDecB, bDecB)
```

[the remark this answers is not in the document]
Comment (algorandskiy) [reply agreement/msgp_gen.go:6872]: looks good, thanks

File: crypto/merklearray/proof.go [hunk 15]
```diff
@@ -20,11 +20,39 @@ import (
...
+// z5
+// z3              z4
+// y       z       z1      z2
+// q   r   s   t   u   v   w   x
+// a b c d e f g h i j k l m n o p
+// ^   ^     ^   ^ ^   ^     ^   ^
+//
+// hints: [b, d, e, g, j, l, m, o]
+// len(hints) = 2^4/2
+// /////////////////////////////////////////////////////////////////////////////////////////
```
Comment (algonautshant) [inline crypto/merklearray/proof.go:55]: The indentations have disfigured my beautiful figures :-(

```suggestion
// Path is bounded by MaxNumLeaves since there could be multiple reveals, and
// given the distribution of the elt positions and the depth of the tree,
// the path length can increase up to 2^MaxTreeDepth / 2
//
// Consider two different reveals for the same tree:
//               z5
//       z3              z4
//   y       z       z1      z2
// q   r   s   t   u   v   w   x
//a b c d e f g h i j k l m n o p
//  ^
// hints: [a, r, z, z4]
// len(hints) = 4
// You need a to combine with b to get q, need r to combine with the computed q and get y, and so on.
//
// The worst case is this:
//                z5
//        z3              z4
//    y       z       z1      z2
//  q   r   s   t   u   v   w   x
// a b c d e f g h i j k l m n o p
// ^   ^     ^   ^ ^   ^     ^   ^
//
// hints: [b, d, e, g, j, l, m, o]
// len(hints) = 2^4/2
// /////////////////////////////////////////////////////////////////////////////////////////
```

File: crypto/merklearray/proof.go [hunk 16]
```diff
@@ -41,12 +42,33 @@ type Proof struct {
...
 
 	Proof
 }
 
+// ProofMaxSizeByElements returns the maximum msgp encoded size of merklearray.Proof structs containing n signatures
+// This is necessary because the allocbounds on the proof are actual theoretical bounds but for calculating maximum
+// proof size for individual message types we have smaller valid bounds.
+// Exported because it's used in the stateproof package for ensuring that SigPartProof constant is correct size.
+func ProofMaxSizeByElements(n int) (s int) {
+	s = 1 + 4
```
[the remark this answers is not in the document]
Comment (iansuvak) [reply crypto/merklearray/proof.go:85]: It's actually not fully manual. It's a modification of the autogenerated test: here https://github.com/algorand/go-algorand/pull/5388/files#diff-add6a5ef46c2629eb7484bf3113df670d70cfdb6661ca69e236a778a7e1cb4c4R334

and we have a test here https://github.com/algorand/go-algorand/pull/5388/files#diff-181a1af1a38a976d2ddce9a94f778520e2ad787079c3cc99dbd1b7483386aa73R248 that ensures that that auto-generated ProofMaxSize and and ProofMaxSizeByElements output the same value when the allocbound is used as the input

File: protocol/tags_test.go [hunk 17]
```diff
@@ -54,6 +56,11 @@ func getConstValues(t *testing.T, fileName string, typeName string) []string {
 			if v.Type == nil || v.Type.(*ast.Ident).Name != typeName {
 				continue
 			}
+
+			if returnNames {
```
Comment (algorandskiy) [inline protocol/tags_test.go:60]: why return early? both usages are
```
	constTags := getConstValues(t, "tags.go", "Tag", false)
	constTags := getConstValues(t, "tags.go", "Tag", true)
```
on the same file so should return hte

File: crypto/merklearray/proof.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (algonautshant) [reply crypto/merklearray/proof.go:85]: The test is very good and reliable. However, the mapping of the function `SingleLeafProofMaxSize` to `SingleLeafProof` is not obvious for someone who is not aware of this arrangement. 

In this particular case, this function is not needed. It can be implemented by setting a constant bound directly to the Proof field in:
```
        Signature struct {
                _struct struct{} `codec:",omitempty,omitemptyarray"`

                Signature             crypto.FalconSignature      `codec:"sig"`
                VectorCommitmentIndex uint64                      `codec:"idx"`
                Proof                 merklearray.SingleLeafProof `codec:"prf"`
```

File: crypto/vrf.go [hunk 18]
```diff
@@ -42,6 +42,11 @@ func init() {
 // VRFVerifier is a deprecated name for VrfPubkey
 type VRFVerifier = VrfPubkey
 
+// VRFVerifierMaxSize forwards to base implementation since it's expected by the msgp generated MaxSize functions
+func VRFVerifierMaxSize() int {
+	return VrfPubkeyMaxSize()
+}
```
Comment (algonautshant) [inline crypto/vrf.go:48]: Why is this needed? The bound can be set like elsewhere in the struct field where this is used.

File: network/msgOfInterest.go [hunk 19]
```diff
@@ -87,3 +87,13 @@ func MarshallMessageOfInterestMap(tagmap map[protocol.Tag]bool) []byte {
 	topics := Topics{Topic{key: "tags", data: []byte(tags)}}
 	return topics.MarshallTopics()
 }
+
+// MessageOfInterestMaxSize returns the maximum size of a MI message sent over the network
+// by encoding all of the tags currenttly in use.
+func MessageOfInterestMaxSize() int {
```
Comment (algonautshant) [inline network/msgOfInterest.go:93]: This should be in the test file, since it is only used from the test.

File: data/basics/units.go [hunk 20]
```diff
@@ -108,6 +108,12 @@ func (a MicroAlgos) MsgIsZero() bool {
 	return a.Raw == 0
 }
 
+// MicroAlgosMaxSize returns maximum possible msgp encoded size of MicroAlgos in bytes.
+// It is expected by msgp generated MaxSize functions
+func MicroAlgosMaxSize() (s int) {
```
Comment (algonautshant) [inline data/basics/units.go:113]: Why is this needed? MicroAlgos has one field of type uint64.

Comment (iansuvak) [reply data/basics/units.go:113]: Same reason as the previous conversations we've had: `msgp` when it sees a type defined outside of the current file it's parsing it assumes that it has this function defined. MicroAlgos is an example of a struct that we generate fully manually and don't use msgp for but we still need this method and the ones above for the rest of the msgp generated functions to work since they call them when they are Marshalling/Unmarshalling themselves or checking their own MaxSize

File: crypto/vrf.go (hunk 18 above)
Comment (algorandskiy) [reply crypto/vrf.go:48]: because of type alias. This type is on TODO for removal.

File: network/msgOfInterest.go (hunk 19 above)
Comment (iansuvak) [reply network/msgOfInterest.go:93]: I can move it but would have to move it to `node_test.go` which seems like an odd place for it and it would break symmetry since all other MaxSize() functions are available in non-test code.

File: crypto/merklearray/proof.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (iansuvak) [reply crypto/merklearray/proof.go:85]: Yeah but we do have a pattern for doing this manual implementation already for basics.MicroAlgos and this solution will continue working if we ever use SingleLeafProof elsewhere. Finding it's `maxtotalbytes` bound and using it is more implicit IMO.

**Review by algorandskiy — APPROVED** [review]

File: ledger/testing/randomAccounts.go (hunk 6 above)
Comment (AlgoAxel) [reply ledger/testing/randomAccounts.go:103]: I know you did not write the conditional, but do you happen to know why we check if the random int is divisible by 5?

File: network/wsPeer.go [hunk 21]
```diff
@@ -530,7 +531,8 @@ func (wp *wsPeer) readLoop() {
 			wp.net.log.Warnf("wsPeer readLoop: received a TS response for a stale request from %s. %d bytes discarded", wp.conn.RemoteAddr().String(), n)
 			continue
 		}
-		slurper.Reset()
+
+		slurper.Reset(uint64(msg.Tag.MaxMessageSize()))
```
Comment (AlgoAxel) [inline network/wsPeer.go:535]: Looks like outside of tests, the slurper is only reset to this one specific size (though maybe I missed something). If that's the case, why not just allow the slurper to set this value at construction and then not reload it?

File: network/wsPeer_test.go [hunk 22]
```diff
@@ -270,7 +270,7 @@ func getProtocolTags(t *testing.T) []string {
 		for _, spec := range genDecl.Specs {
 			if valueSpec, ok := spec.(*ast.ValueSpec); ok {
 				for _, n := range valueSpec.Names {
-					if n.Name == "TagLength" {
+					if strings.HasSuffix(n.Name, "MaxSize") || n.Name == "TagLength" {
 						continue
 					}
 					declaredTags = append(declaredTags, n.Name)
```
Comment (AlgoAxel) [inline network/wsPeer_test.go:276]: if I'm reading this block right, we're trying to iterate over the valueSpec of a given tag, and through excluding MaxSize and TagLength, we expect that the only remaining thing is the declared tag's name? This seems fragile if/when more is added to the spec. Since it's just a test I don't know that it's worth pursuing.

File: protocol/tags.go [hunk 23]
```diff
@@ -46,6 +46,90 @@ const (
...
+
+// UniEnsBlockReqTagMaxSize is the maximum size of a UniEnsBlockReqTag message
+const UniEnsBlockReqTagMaxSize = 67
+
+// VoteBundleTagMaxSize is the maximum size of a VoteBundleTag message
+// Matches current network.MaxMessageLength
+const VoteBundleTagMaxSize = 6 * 1024 * 1024
+
+// MaxMessageSize returns the maximum size of a message for a given tag
+func (tag Tag) MaxMessageSize() uint64 {
```
Comment (AlgoAxel) [inline protocol/tags.go:100]: Instead of a manually maintained constants with switch statement, why not a map?

```
MaxMessageSizes := map[tag]Value {
VoteBundle: 6 * 1024 * 1024,
...
}
...

func MaxMessageSize() {
return MaxMessageSizes[tag.Name]
}
```
(awful pseudocode, but you get the idea)

**Review by AlgoAxel — COMMENTED** [review]
I reviewed primarily the Network and Tags modules, and looked lightly at the rest.

File: protocol/tags.go (hunk 23 above)
Comment (algorandskiy) [reply protocol/tags.go:100]: this switch appears ~15% faster

File: ledger/testing/randomAccounts.go (hunk 6 above)
Comment (iansuvak) [reply ledger/testing/randomAccounts.go:103]: It seems like we want only 4/5s of generated AssetParams to have these fields defined? I don't know though

File: network/wsPeer.go (hunk 21 above)
Comment (iansuvak) [reply network/wsPeer.go:535]: This is a for loop that runs for the lifetime of the peer. Each iteration of the for loop will have a different msg and therefore a different Tag and it's `MaxMessageSize()`

File: network/wsPeer_test.go (hunk 22 above)
Comment (iansuvak) [reply network/wsPeer_test.go:276]: Yeah it is fragile as you can see since both myself and Shant had to add exclusions here for MaxSize and TagLength respectively. Hopefully we won't be adding a lot more specs to protocol.Tags

**Review by algorandskiy — APPROVED** [review]

**Review by AlgoAxel — APPROVED** [review]
